### ISSUE TITLE
automatic machine provisioning

### ISSUE THREAD

**Issue by smira** [siderolabs/omni#270]
## Provisioning

1. Omni creates a `MachineRequest` to provision a `Machine` in some cloud with some settings (region, size, ...)
2. Omni Cloud Provider (a plugin to Omni) picks up the `MachineRequest`, and provisions the machine in the cloud.
3. Omni Cloud Provider creates `MachineRequestStatus` describing the future machine which will join Omni (e.g. it has the UUID of the Machine that is going to be created/going to join Omni).
4. As a provisioned `Machine` joins Omni, provider matches it with the request and labels/annotates it as requested (for example, label indicates that this Machine should be part of a `MachineClass` and it will join the `MachineSet`).

## De-provisioning

1. Machine is removed in the cloud --> Omni Cloud Provider picks it up and starts removing `Machine` from Omni.
2. MachineRequest is removed -> Cloud Provider removes the `Machine`, removes machine in the cloud, cleans up everything.
3. `Machine` is removed from Omni (e.g. faulty hardware) -> Omni Cloud Provider cleans it in the cloud.

**Comment by smira** [comment]
## Omni Cloud Provider Deployment

1. Might run with Omni (customer cloud access keys are given to Omni)
2. Might run outside of Omni (auths to Omni using a service account, access keys are within external deployment)

**Comment by smira** [comment]
# Omni Cloud Provider 

1. Connects to Omni API (some specific role, like 'Cloud Provider').
2. Fetches cloud provider config resource from Omni (or uses config supplied via the environment).
3. Watches `MachineRequest` resources with specific label (targeting this provider ID).
4. If this is a new `MachineRequest` that needs to be satisfied:
   * puts a finalizer on the `MachineRequest`
   * creates a `MachineRequestStatus` in not ready state, uses that to store internal progress status
   * starts provisioning a machine in the cloud, it might update `MachineRequestStatus` with some data to restart operations on failure
   * once the machine is ready (booting), update `MachineRequestStatus` to be ready and put the UUID of the machine which is going to join Omni (or labels the machine in a unique way); Omni should be able to match a joining `Machine` back to the `MachineRequest`
5. If this is a tearing down `MachineRequest`:
  * destroy the infrastructure
  * destroy the `MachineRequestStatus` resource
  * remove the finalizer on `MachineRequest`

**Comment by smira** [comment]
## `MachineRequest` Resource

```yaml
metadata:
   id: machine-xyz
spec:
```

**Comment by smira** [comment]
## Machine operations

```yaml
metadata:
   # Omni-managed
   id: machine-xyz
   type: MachineControlSpec
spec:
   poweredOn: true/false # current status
   rebootID: 33 # need better name, if changes, new reboot is executed
---
metadata:
   # Cloud provider-managed
   id: machine-xyz
   type: MachineControlStatus
spec:
     poweredOn: true/false # current status
     lastRebootID: 33 # if matches rebootID, the last reboot got executed
```

**Comment by Unix4ever** [comment]
We might additionally want to:
1. Delete the link for the tearingdown `MachineRequestStatus`
2. Do we want to delete `MachineRequest` if the link is delete? Or do we want to somehow indicate that the cloud provider should re-create the machine? Or do we want to not allow deleting links which are created by `MachineRequest` and make the UI delete the machine request instead?

**Comment by utkuozdemir** [comment]
Moving this to done, as most of the basic blocks are done - we can do the machine power status / reboot etc in a separate ticket.
